html2text.1
.\" manpage for html2text
.\"
.TH html2text 1 2024\-10\-29
.SH NAME
html2text \- an advanced HTML\-to\-text converter
.SH SYNOPSIS
.B html2text -help
.br
.B html2text -version
.br
.B html2text
[
.B \-check
] [
.B \-debug\-scanner
] [
.B \-debug\-parser
] [
.B \-rcfile
.I path
] [
.B \-width
.I width
] [
.B \-o
.I output-file
] [
.B \-nobs
] [
.B \-from_encoding
.I encoding
] [
.B \-to_encoding
.I encoding
] [
.B \-ascii
] [
.B \-utf8
] [
.IR input-file " ..."
]
.SH DESCRIPTION
.B html2text
reads HTML documents from the
.IR input-file s,
formats each of them into a stream of UTF-8 encoded characters,
and writes the result to standard output (or into
.IR output-file ,
if the
.B -o
command line option is used).
.P
If no
.IR input-file s
are specified on the command line,
.B html2text
reads from standard input. A dash as the
.I input-file
is an alternate way to specify standard input.
.P
.B html2text
understands all HTML 3.2 constructs, but can render only part of them due to
the limitations of the text output format. However, the program attempts to
provide good substitutes for the elements it cannot render.
.B html2text
parses HTML 4 input, too, but not always as successful as other HTML
processors. It also accepts syntactically incorrect input, and attempts to
interpret it "reasonably".
.P
The way
.B html2text
formats the HTML documents is controlled by formatting properties read
from an RC file.
.B html2text
attempts to read
.I $HOME/.html2textrc
(or the file specified by the
.B -rcfile
command line option); if that file cannot be read,
.B html2text
attempts to read
.IR /etc/html2textrc .
If no RC file can be read (or if the RC file does not override all
formatting properties), then "reasonable" defaults are assumed. The
RC file format is described in the
.BR html2textrc (5)
manual page.
.SH OPTIONS
.TP
.B \-ascii
By default,
.B html2text
uses
.B UTF-8
for its output. Specifying this option, plain
.B ASCII
is used instead. Non\-ASCII characters are rendered via iconv's
transliteration feature.  As such this option is an alias for
.B -to_encoding
.IR ASCII//TRANSLIT .
.TP
.B \-utf8
Assume both terminal and input stream are in
.B UTF-8
mode. This is an alias for
.B -from_encoding
.IR UTF-8
.B -to_encoding
.IR UTF-8 .
.TP
.BI "\-from_encoding " encoding
Sets the encoding of the input file or stream to the given encoding.  By
default,
.B html2text
tries to obtain the encoding from the input file and uses ISO-8859-1
encoding as fallback.  You might want to override the detection or
fallback using this option.  See 
.B iconv -l
for a list of supported encodings.
.TP
.BI "\-to_encoding " encoding
Use the given encoding while writing output, instead of UTF-8.  The same
set of encodings as for the
.B \-from_encoding
option can be used.
.TP
.B \-check
This option is for diagnostic purposes: The HTML document is only parsed and
not processed otherwise. In this mode of operation,
.B html2text
will report on parse errors and scan errors, which it does not in other modes
of operation. Note that parse and scan errors are not fatal for
.BR html2text ,
but may cause mis-interpretation of the HTML code and/or portions of the
document being swallowed.
.TP
.B \-debug\-parser
Let
.B html2text
report on the tokens being shifted, rules being applied, etc., while scanning
the HTML document. This option is for diagnostic purposes.
.TP
.B \-debug\-scanner
Let
.B html2text
report on each lexical token scanned, while scanning the HTML document. This
option is for diagnostic purposes.
.TP
.B \-help
Print command line summary and exit.
.TP
.B \-nobs
By default,
.B html2text
renders underlined letters with sequences like "underscore-backspace-character"
and boldface letters like "character-backspace-character", which works fine
when the output is piped into
.BR more (1),
.BR less (1),
or similar. For other applications, or when redirecting the output into a file,
it may be desirable not to render character attributes with such backspace
sequences, which can be accomplished with this command line option.
.TP
.BI \-o " output\-file"
Write the output to
.I output\-file
instead of standard output. A dash as the
.I output\-file
is an alternate way to specify the standard output.
.TP
.BI \-rcfile " path"
Attempt to read the file specified in
.I path
as RC file.
.TP
.B \-links
Tags all external links with a number between brackets
.RI "([" num "]),"
and produces a numbered list at the end of the document with all link
targets.
.TP
.B \-version
Print program version and exit.
.TP
.BI \-width " width"
By default,
.B html2text
formats the HTML documents for a screen width of 79 characters. If redirecting
the output into a file, or if your terminal has a width other than 80
characters, or if you just want to get an idea how
.B html2text
deals with large tables and different terminal widths, you may want to specify
a different
.IR width .
.SH FILES
.TP
.I /etc/html2textrc
System wide parser configuration file.
.TP
.I $HOME/.html2textrc
Personal parser configuration file, overrides the system wide values.
.SH "CONFORMING TO"
.B "HTML 3.2"
(HTML 3.2 Reference Specification \- http://www.w3.org/TR/REC-html32).
.SH RESTRICTIONS
.B html2text
was written to convert HTML 3.2 documents. When using it with HTML 4 or
even XHTML 1 documents, some constructs present only in these HTML versions
might not be rendered.
.SH AUTHOR
.B html2text
was written up to version 1.2.2 by Arno Unkrig <arno@unkrig.de>
for GMRS Software GmbH, Unterschleissheim.
.P
Up to version 1.3.2a, the maintainer was
Martin Bayer <mbayer@zedat.fu-berlin.de>.
.br
.P
This version of
.B html2text
comes from
https://gitlab.com/grobian/html2text.  Please use the GitLab page to
file issues and improvements.
.SH SEE ALSO
.BR html2textrc (5),
.BR less (1),
.BR more (1)


ChangeLog.md
# 2.5.0 (unreleased master branch)

### New Features

### Bugfixes

* [Issue #69](https://gitlab.com/grobian/html2text/-/issues/69)
  Fix version reporting in -version and -help output.
* [Issue #70](https://gitlab.com/grobian/html2text/-/issues/70)
  Fix crash when reading multiple files.


# 2.4.0 (02-08-2025)

### New Features

* [Issue #18](https://gitlab.com/grobian/html2text/-/issues/18)
  Allow retrieving remote URLs directly.

### Bugfixes

* Fixed various rendering issues in tables.


# 2.3.0 (06-06-2025)

### New Features

* [Issue #62](https://gitlab.com/grobian/html2text/-/issues/62)
  use ANSI-escapes instead of backspace sequences, this includes some
  colour handling

### Bugfixes

* [Issue #59](https://github.com/grobian/html2text/issues/59)
  Fix build for C++ compilers that do not default to C++11 or more
  recent.
* [Issue #61](https://https://gitlab.com/grobian/html2text/-/issues/61)
  Fix handling of malformed attributes in tags
* [PR #54](https://gitlab.com/grobian/html2text/-/merge_requests/54)
  Fix unput type checkbox rendering
* [Issue #64](https://https://gitlab.com/grobian/html2text/-/issues/64)
  Don't write rendering sequences to output files by default
* [Issue #66](https://gitlab.com/grobian/html2text/-/issues/66)
  Emit italics escape code for &lt;i&gt;


# 2.2.3 (31-07-2023)

### New Features

### Bugfixes

* [Issue #57](https://github.com/grobian/html2text/issues/57)
  de-escape link hrefs so they can be copy-pasted in a browser
* [Issue #56](https://github.com/grobian/html2text/issues/56)
  drop usage of sprintf from the codebase


# 2.2.2 (30-07-2023)

### New Features

### Bugfixes

* Fix building on platforms requiring libiconv


# 2.2.1 (30-07-2023)

### New Features

### Bugfixes

* Fix release script to produce working tarball


# 2.2.0 

### New Features

* [Issue #52](https://github.com/grobian/html2text/issues/52)
  Support -utf8 flag as alias for compatibility with <2.0.0 releases
* Build-system now uses autotools-based configure and Makefiles
* Suppress xml namespace tags in output (<?xml ...)
* [Issue #12](https://github.com/grobian/html2text/issues/12)
  [Issue #26](https://github.com/grobian/html2text/issues/26)
  Whitespace normalisation now works over tags, removing extra
  and adding missing where appropriate
* Ignore &zwnj; to improve monospace rendering

### Bugfixes

* [PR #48](https://github.com/grobian/html2text/pull/48)
  Fix typos in manpage


html2text.cpp
/*
 * Portions Copyright (c) 1999 GMRS Software GmbH
 * Carl-von-Linde-Str. 38, D-85716 Unterschleissheim, http://www.gmrs.de
 * All rights reserved.
 *
 * Author: Arno Unkrig <arno@unkrig.de>
 *
 * Copyright 2020-2025 Fabian Groffen <grobian@gentoo.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License in the file COPYING for more details.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <iostream>
#include <string.h>
#include <stdlib.h>

#include "html.h"
#include "HTMLControl.h"
#include "HTMLDriver.h"
#include "iconvstream.h"
#include "format.h"


static const char *usage = "\
Usage:\n\
  html2text -help\n\
  html2text -version\n\
  html2text [ -check ] [ -debug-scanner ] [ -debug-parser ] \\\n\
     [ -rcfile <file> ] [ -width <w> ] [ -nobs ] [ -bs ] [ -links ]\\\n\
     [ -from_encoding ] [ -to_encoding ] [ -ascii ]\\\n\
     [ -o <file> ] [ <input-file> ] ...\n\
Formats HTML document(s) read from <input-file> or STDIN and generates ASCII\n\
text.\n\
  -help          Print this text and exit\n\
  -version       Print program version and copyright notice\n\
  -check         Do syntax checking only\n\
  -debug-scanner Report parsed tokens on STDERR (debugging)\n\
  -debug-parser  Report parser activity on STDERR (debugging)\n\
  -rcfile <file> Read <file> instead of \"$HOME/.html2textrc\"\n\
  -width <w>     Optimize for screen widths other than 79\n\
  -nobs          Do not render boldface, underlining, colours, etc.\n\
  -bs            Render boldface and underlining using backspaces\n\
  -ansi          Render bold, underline and colours in output (default)\n\
  -links         Generate reference list with link targets\n\
  -from_encoding Treat input encoded as given encoding\n\
  -to_encoding   Output using given encoding\n\
  -ascii         Use plain ASCII for output instead of UTF-8\n\
                 alias for: -to_encoding ASCII//TRANSLIT \n\
  -utf8          Assume both terminal and input stream are in UTF-8 mode\n\
                 alias for: -from_encoding UTF-8 -to_encoding UTF-8 \n\
  -o <file>      Redirect output into <file>\n\
";

int
main(int argc, char **argv)
{
	if (argc == 2 && !strcmp(argv[1], "-help")) {
		std::cout
			<< "This is html2text, version " VERSION << std::endl
			<< std::endl
			<< usage;
		exit(0);
	}

	if (argc == 2 && !strcmp(argv[1], "-version")) {
		std::cout
			<< "This is html2text, version " VERSION << std::endl
			<< std::endl
			<< "The latest version can be found at "
			<< "https://gitlab.com/grobian/html2text" << std::endl
			<< std::endl
			<< "This program is distributed in the hope that it will "
			<< "be useful, but WITHOUT" << std::endl
			<< "ANY WARRANTY; without even the implied warranty of "
			<< "MERCHANTABILITY or FITNESS" << std::endl
			<< "FOR A PARTICULAR PURPOSE. See the GNU General Public "
			<< "License for more details." << std::endl
			<< std::endl;
		exit(0);
	}

	int mode = HTMLDriver::PRINT_AS_ASCII;
	bool debug_scanner = false;
	bool debug_parser = false;
	const char *home = getenv("HOME");
	const char *rcfile = NULL;
	int width = 79;
	const char *output_file_name = "-";
	bool use_backspaces = false;
	bool use_ansi_escapes = true;
	bool explicit_rendering_req = false;
	bool enable_links = false;
	const char *from_encoding = NULL;
	const char *to_encoding = NULL;
	const char *widthstr = NULL;
	const char **extarg = NULL;

	int i;
	for (i = 1; i < argc && argv[i][0] == '-' && argv[i][1]; i++) {
		const char *arg = argv[i];

		if (!strcmp(arg, "-check")) {
			mode = HTMLDriver::SYNTAX_CHECK;
		} else if (!strcmp(arg, "-debug-scanner")) {
			debug_scanner = true;
		} else if (!strcmp(arg, "-debug-parser")) {
			debug_parser = true;
		} else if (!strcmp(arg, "-rcfile")) {
			extarg = &rcfile;
		} else if (!strcmp(arg, "-links")) {
			enable_links = true;
		} else if (!strcmp(arg, "-width")) {
			extarg = &widthstr;
		} else if (!strcmp(arg, "-o")) {
			extarg = &output_file_name;
		} else if (!strcmp(arg, "-nobs")) {
			use_backspaces = false;
			use_ansi_escapes = false;
			explicit_rendering_req = true;
		} else if (!strcmp(arg, "-bs")) {
			use_backspaces = true;
			use_ansi_escapes = false;
			explicit_rendering_req = true;
		} else if (!strcmp(arg, "-ansi")) {
			use_backspaces = false;
			use_ansi_escapes = true;
			explicit_rendering_req = true;
		} else if (!strcmp(arg, "-from_encoding")) {
			extarg = &from_encoding;
		} else if (!strcmp(arg, "-to_encoding")) {
			extarg = &to_encoding;
		} else if (!strcmp(arg, "-ascii")) {
			to_encoding = "ASCII//TRANSLIT";  /* create things like (c) */
		} else if (!strcmp(arg, "-utf8")) {
			from_encoding = "UTF-8";
			to_encoding = "UTF-8";
		} else {
			std::cerr
				<< "Unrecognized command line option \""
				<< arg
				<< "\", try \"-help\"."
				<< std::endl;
			exit(1);
		}

		if (extarg != NULL) {
			if (++i >= argc) {
				std::cerr
					<< "Option \"" << arg << "\" needs an additional "
					<< "argument."
					<< std::endl;
				exit(1);
			}
			*extarg = argv[i];

			/* handle the only integer argument inline */
			if (extarg == &widthstr) {
				int nwidth = atoi(widthstr);
				if (nwidth > 10) {
					width = nwidth;
				} else {
					std::cerr
						<< "width '" << nwidth << "' invalid, must be >10"
						<< std::endl;
					exit(1);
				}
			}
			extarg = NULL;
		}
	}
	if (i > argc) {
		std::cerr
			<< "Error: Required parameter after \""
			<< argv[argc - 1]
			<< "\" missing."
			<< std::endl;
		exit(1);
	}

	/* historical default used to be ISO-8859-1, auto is not a valid
	 * encoding, but handled in iconvstream */
	if (from_encoding == NULL)
		from_encoding = "auto";

	/* this is probably the output we want on 99% of all terminals */
	if (to_encoding == NULL)
		to_encoding = "UTF-8";

	const char *const *input_files;
	int number_of_input_files;

	if (i >= argc) {
		static const char *const x = "-";
		input_files = &x;
		number_of_input_files = 1;
	} else {
		input_files = argv + i;
		number_of_input_files = argc - i;
	}

	{
		std::ifstream ifs;
		std::string homerc;

		if (rcfile == NULL && home != NULL) {
			homerc = string(home) + "/.html2textrc";
			rcfile = homerc.c_str();
		}
		if (rcfile != NULL)
			ifs.open(rcfile);
		if (rcfile == NULL || !ifs.rdbuf()->is_open())
			ifs.open("/etc/html2textrc");
		if (ifs.rdbuf()->is_open()) {
			Formatting::loadProperties(ifs);
			ifs.close();
		}
	}

	iconvstream is;

	is.open_os(output_file_name, to_encoding);
	if (!is.os_open()) {
		std::cerr
			<< "Could not open output file \""
			<< output_file_name
			<< "\": "
			<< is.open_error_msg()
			<< std::endl;
		exit(1);
	}

	if (!is.os_isatty() && !explicit_rendering_req) {
		/* disable "weird" chars unless explicitly requested, #64 */
		use_backspaces = false;
		use_ansi_escapes = false;
	}

	/*
	 * Set up printing.
	 */
	Area::use_backspaces = use_backspaces;
	Area::use_ansi       = use_ansi_escapes;

	for (i = 0; i < number_of_input_files; ++i) {
		const char *input_file = input_files[i];

		if (number_of_input_files != 1)
			is << "###### " << input_file << " ######" << endl;

		is.open_is(input_file, from_encoding);
		if (!is.is_open()) {
			std::cerr
				<< "Opening input file \""
				<< input_file
				<< "\": "
				<< is.open_error_msg()
				<< std::endl;
			exit(1);
		}

		HTMLControl control(is, mode, debug_scanner, input_file);
		HTMLDriver driver(control, is, enable_links,
				width, mode, debug_parser);

		if (driver.parse() != 0)
			exit(1);
	}

	is.close();

	return 0;
}
